File: frontend/src/stores/slack.ts
import { ref } from 'vue'
import { defineStore } from 'pinia'
import { api } from '@/api/client'
import { useAppStore } from './app'

export interface SlackConfig {
  id: number
  token: boolean
  cookie: boolean
  cred_exists: boolean
  channels: string[]
  collect_interval: number
  collect_lookback: number
  last_collect_at: string | null
  total_messages: number
  last_refresh_at: string | null
  updated_at: string
  created_at: string
}

export interface SlackStatus {
  cred_exists: boolean
  cred_valid: boolean
  last_collect_at: string | null
  total_messages: number
  collect_interval: number
  collect_lookback: number
  last_refresh_at: string | null
}

export interface SlackChannel {
  id: string
  name: string
  display: string
  num_members: number
  topic: string
  purpose: string
}

export const useSlackStore = defineStore('slack', () => {
  const config = ref<SlackConfig | null>(null)
  const status = ref<SlackStatus | null>(null)
  const showManager = ref(false)
  const configLoading = ref(false)
  const availableChannels = ref<SlackChannel[]>([])
  const channelsLoading = ref(false)
  const tokenInput = ref('')
  const cookieInput = ref('')
  const newChannel = ref('')
  const collectInterval = ref(360)
  const collectLookback = ref(1440)
  const saving = ref(false)
  const collecting = ref(false)
  const testing = ref(false)
  const refreshing = ref(false)

  async function loadConfig() {
    configLoading.value = true
    try {
      const data: any = await api('/api/slack/config')
      config.value = data as SlackConfig
      collectInterval.value = data.collect_interval || 360
      collectLookback.value = data.collect_lookback || 1440
      if (data.token) tokenInput.value = data.token
      if (data.cookie) cookieInput.value = data.cookie
    } catch (e: any) {
      useAppStore().showToast('加载 Slack 配置失败，尝试自动刷新凭证...', e.message, 'error')
      await refreshToken()
    } finally {
      configLoading.value = false
    }
  }

  async function loadStatus() {
    try {
      const data: any = await api('/api/slack/status')
      status.value = data as SlackStatus
    } catch (e: any) {
      // silent
    }
  }

  async function loadAvailableChannels() {
    if (channelsLoading.value) return
    channelsLoading.value = true
    try {
      const data: any = await api('/api/slack/channels')
      availableChannels.value = (data.channels || []) as SlackChannel[]
    } catch (e: any) {
      useAppStore().showToast('加载频道列表失败', e.message, 'error')
    } finally {
      channelsLoading.value = false
    }
  }

  async function testAuth() {
    const token = tokenInput.value.trim()
    const cookie = cookieInput.value.trim()
    if (!token || !cookie) {
      useAppStore().showToast('请先填写 token 和 cookie', '', 'error')
      return
    }
    testing.value = true
    try {
      const data: any = await api('/api/slack/test-auth', {
        method: 'POST',
        body: JSON.stringify({ token, cookie }),
      })
      if (data.ok) {
        useAppStore().showToast('凭证有效', `用户: ${data.user}，团队: ${data.team}`, 'success')
        await loadStatus()
      } else {
        useAppStore().showToast('凭证无效', data.error || 'unknown', 'error')
      }
    } catch (e: any) {
      useAppStore().showToast('测试失败', e.message, 'error')
    } finally {
      testing.value = false
    }
  }

  async function saveConfig() {
    if (saving.value) return
    saving.value = true
    try {
      const payload: any = {
        channels: config.value?.channels || [],
        collect_interval: collectInterval.value,
        collect_lookback: collectLookback.value,
      }
      if (tokenInput.value.trim()) payload.token = tokenInput.value.trim()
      if (cookieInput.value.trim()) payload.cookie = cookieInput.value.trim()

      const data: any = await api('/api/slack/config', {
        method: 'PUT',
        body: JSON.stringify(payload),
      })
      config.value = data as SlackConfig
      if (data.token) tokenInput.value = data.token
      if (data.cookie) cookieInput.value = data.cookie
      useAppStore().showToast('Slack 配置已保存', '', 'success')
    } catch (e: any) {
      useAppStore().showToast('保存 Slack 配置失败', e.message, 'error')
    } finally {
      saving.value = false
    }
  }

  async function addChannel() {
    const channel = newChannel.value.trim()
    if (!channel) {
      useAppStore().showToast('请选择或输入频道', '', 'error')
      return
    }
    if (!channel.startsWith('#')) {
      useAppStore().showToast('频道名称必须以 # 开头', '', 'error')
      return
    }
    if (config.value?.channels?.includes(channel)) {
      useAppStore().showToast('频道已存在', channel, 'warning')
      return
    }
    try {
      const data: any = await api('/api/slack/config/channels', {
        method: 'POST',
        body: JSON.stringify({ channel }),
      })
      config.value = data as SlackConfig
      newChannel.value = ''
      useAppStore().showToast('频道已添加', channel, 'success')
    } catch (e: any) {
      useAppStore().showToast('添加频道失败', e.message, 'error')
    }
  }

  async function removeChannel(channel: string) {
    const result = await useAppStore().showConfirm({
      title: '移除频道',
      message: `确认移除频道 ${channel}？`,
      confirmText: '移除',
      danger: true,
    })
    if (!result.confirmed) return
    try {
      const data: any = await api(`/api/slack/config/channels/${encodeURIComponent(channel)}`, {
        method: 'DELETE',
      })
      config.value = data as SlackConfig
      useAppStore().showToast('频道已移除', channel, 'info')
    } catch (e: any) {
      useAppStore().showToast('移除频道失败', e.message, 'error')
    }
  }

  async function triggerCollect() {
    if (collecting.value) return
    collecting.value = true
    try {
      const data: any = await api('/api/slack/collect', { method: 'POST' })
      useAppStore().showToast('Slack 采集完成', `获取 ${data.total_fetched || 0} 条，新增 ${data.total_stored || 0} 条`, 'success')
      await loadStatus()
      await loadConfig()
    } catch (e: any) {
      useAppStore().showToast('Slack 采集失败', e.message, 'error')
    } finally {
      collecting.value = false
    }
  }

  async function clearData() {
    const result = await useAppStore().showConfirm({
      title: '清除 Slack 数据',
      message: '确认清除所有 Slack 采集数据？此操作不可恢复。',
      confirmText: '清除',
      danger: true,
    })
    if (!result.confirmed) return
    try {
      const data: any = await api('/api/slack/data', { method: 'DELETE' })
      useAppStore().showToast('Slack 数据已清除', `删除了 ${data.deleted} 条记录`, 'info')
      await loadStatus()
    } catch (e: any) {
      useAppStore().showToast('清除失败', e.message, 'error')
    }
  }

  async function refreshToken() {
    refreshing.value = true
    try {
      const data: any = await api('/api/slack/refresh-token', { method: 'POST' })
      if (data.ok) {
        config.value = data as SlackConfig
        if (data.token) tokenInput.value = data.token
        if (data.cookie) cookieInput.value = data.cookie
        useAppStore().showToast('凭证已刷新', '', 'success')
        await loadStatus()
      } else {
        useAppStore().showToast('刷新失败', data.error || '请检查 .env 中 SLACK_EMAIL / SLACK_PASSWORD', 'error')
      }
    } catch (e: any) {
      useAppStore().showToast('刷新失败', e.message, 'error')
    } finally {
      refreshing.value = false
    }
  }

  function getAvailableNotConfigured(): SlackChannel[] {
    const configured = config.value?.channels || []
    return availableChannels.value.filter(ch => !configured.includes(`#${ch.name}`))
  }

  function openManager() {
    showManager.value = true
    loadConfig()
    loadStatus()
    loadAvailableChannels()
  }

  function closeManager() {
    showManager.value = false
  }

  return {
    config, status, showManager, configLoading, availableChannels, channelsLoading,
    tokenInput, cookieInput, newChannel, collectInterval, collectLookback, saving, collecting, testing, refreshing,
    loadConfig, loadStatus, loadAvailableChannels,
    testAuth, saveConfig, addChannel, removeChannel,
    triggerCollect, clearData, refreshToken, getAvailableNotConfigured,
    openManager, closeManager,
  }
})

File: app/api/slack.py
"""
Slack API - Slack 采集配置管理
"""
import json
import logging
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import AIMemory, SlackConfig, _iso_utc

logger = logging.getLogger(__name__)
router = APIRouter()


class SlackConfigUpdate(BaseModel):
    token: str = ""
    cookie: str = ""
    channels: list[str] = []
    collect_interval: int = 360
    collect_lookback: int = 1440


def _utcnow() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _get_or_create_config(db: Session) -> SlackConfig:
    config = db.query(SlackConfig).first()
    if not config:
        config = SlackConfig(
            channels=json.dumps([]),
            collect_interval=360,
            created_at=_utcnow(),
        )
        db.add(config)
        db.commit()
        db.refresh(config)
    return config


@router.get("/config")
async def get_slack_config(db: Session = Depends(get_db)):
    """获取 Slack 配置"""
    config = _get_or_create_config(db)
    return config.to_dict()


@router.put("/config")
async def update_slack_config(req: SlackConfigUpdate, db: Session = Depends(get_db)):
    """更新 Slack 配置（含 token/cookie）"""
    config = _get_or_create_config(db)
    if req.token:
        config.token = req.token
    if req.cookie:
        config.cookie = req.cookie
    config.channels = json.dumps(req.channels)
    config.collect_interval = req.collect_interval
    config.collect_lookback = req.collect_lookback or 1440
    config.cred_exists = bool(req.token and req.cookie)
    config.updated_at = _utcnow()
    db.commit()
    db.refresh(config)
    return config.to_dict()


@router.post("/config/channels")
async def add_channel(channel_data: dict, db: Session = Depends(get_db)):
    """新增一个频道"""
    channel = channel_data.get("channel", "").strip()
    if not channel:
        raise HTTPException(status_code=400, detail="channel is required")
    if not channel.startswith("#"):
        raise HTTPException(status_code=400, detail="channel must start with #")

    config = _get_or_create_config(db)
    channels = json.loads(config.channels) if isinstance(config.channels, str) else (config.channels or [])
    if channel in channels:
        raise HTTPException(status_code=400, detail=f"channel {channel} already exists")
    channels.append(channel)
    config.channels = json.dumps(channels)
    config.updated_at = _utcnow()
    db.commit()
    return config.to_dict()


@router.delete("/config/channels/{channel:path}")
async def delete_channel(channel: str, db: Session = Depends(get_db)):
    """删除一个频道"""
    config = _get_or_create_config(db)
    channels = json.loads(config.channels) if isinstance(config.channels, str) else (config.channels or [])
    if channel not in channels:
        raise HTTPException(status_code=404, detail=f"channel {channel} not found")
    channels.remove(channel)
    config.channels = json.dumps(channels)
    config.updated_at = _utcnow()
    db.commit()
    return config.to_dict()


@router.get("/status")
async def get_slack_status(db: Session = Depends(get_db)):
    """获取 Slack 采集状态"""
    from app.services.slack_collector import SlackCollector
    collector = SlackCollector()
    cred_exists = collector.has_credentials()
    config = _get_or_create_config(db)
    total = db.query(AIMemory).filter(AIMemory.source_type == "slack").count()
    cred_valid = collector.check_credential() if cred_exists else False
    return {
        "cred_exists": cred_exists or config.cred_exists or False,
        "cred_valid": cred_valid,
        "last_collect_at": _iso_utc(config.last_collect_at),
        "total_messages": total,
        "collect_interval": config.collect_interval or 360,
        "collect_lookback": config.collect_lookback or 1440,
        "last_refresh_at": _iso_utc(config.last_refresh_at),
    }


@router.get("/channels")
async def list_slack_channels():
    """获取 Slack 工作区所有可见频道列表"""
    from app.services.slack_collector import SlackCollector
    collector = SlackCollector()
    channels = collector.list_channels()
    return {"channels": channels}


@router.post("/collect")
async def trigger_collect(db: Session = Depends(get_db)):
    """手动触发一次采集"""
    from app.services.slack_collector import SlackCollector
    collector = SlackCollector()
    stats = collector.collect()
    return stats


@router.post("/refresh-token")
async def refresh_slack_token(db: Session = Depends(get_db)):
    """手动触发凭证刷新"""
    from app.services.slack_collector import SlackCollector
    collector = SlackCollector()
    ok = collector._refresh_credentials()
    if ok:
        config = _get_or_create_config(db)
        return {"ok": True, **config.to_dict()}
    return {"ok": False, "error": "refresh failed, check SLACK_EMAIL / SLACK_PASSWORD in .env"}


@router.post("/test-auth")
async def test_auth(req: SlackConfigUpdate, db: Session = Depends(get_db)):
    """测试凭证是否有效"""
    import requests, time
    token = req.token or ""
    cookie = req.cookie or ""
    if not token or not cookie:
        raise HTTPException(status_code=400, detail="token and cookie are required")
    s = requests.Session()
    s.headers.update({"Authorization": f"Bearer {token}"})
    s.headers.update({"Cookie": f"d={cookie}; d-s={str(int(time.time()) - 10)}"})
    try:
        resp = s.get("https://slack.com/api/auth.test", timeout=15)
        data = resp.json()
        if data.get("ok"):
            return {"ok": True, "user": data.get("user"), "team": data.get("team")}
        return {"ok": False, "error": data.get("error", "unknown")}
    except Exception as e:
        return {"ok": False, "error": str(e)}


@router.delete("/data")
async def clear_slack_data(db: Session = Depends(get_db)):
    """清除所有 Slack 知识库数据"""
    deleted = db.query(AIMemory).filter(
        AIMemory.source_type == "slack"
    ).delete(synchronize_session=False)
    config = _get_or_create_config(db)
    config.total_messages = 0
    db.commit()
    return {"deleted": deleted}